src/h5web/visualizations/heatmap/HeatmapVis.tsx

```diff
@@ -21,29 +24,42 @@ function HeatmapVis(props: Props): JSX.Element {
   const values = dataArray.data as number[];
 
   const {
-    dataDomain,
     customDomain,
     scaleType,
     colorMap,
     keepAspectRatio,
     showGrid,
-    initDataDomain,
+    resetDomains,
   } = useHeatmapConfig();
 
-  // width / height <=> cols / rows
-  const aspectRatio = keepAspectRatio ? cols / rows : undefined;
+  const dataDomain = useMemo(() => findDomain(values), [values]);
+  const prevDataDomain = usePrevious(dataDomain);
+
+  // If `dataDomain` just changed, use `undefined` custom domain for this render until config is updated through `useEffect` below
+  const freshCustomDomain =
+    dataDomain !== prevDataDomain ? undefined : customDomain;
 
   useEffect(() => {
-    initDataDomain(values);
-  }, [initDataDomain, values]);
+    // When `dataDomain` changes:
+    // - set data domain in config so toolbar can render domain slider with correct bounds;
+    // - reset custom domain in config so toolbar can reset value of domain slider, and so Heatmap can re-render with correct, `undefined` custom domain.
+    resetDomains(dataDomain);
+  }, [dataDomain, resetDomains]);
```

Comment:
And the third part: we update the config, which update the domain slider and the toolbar and re-renders the Heatmap vis. During this second render, `dataDomain !== prevDataDomain` is `false`, so we use the `customDomain` from the config... but it's fine because `customDomain` is now `undefined` in the config and therefore no longer stale.